packages/eslint-plugin/src/rules/no-circular-import.ts

```diff
@@ -0,0 +1,181 @@
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+type Options = [];
+type MessageIds = 'noCircularImport';
+
+interface Edge {
+  filename: string;
+  specifier: string;
+}
+
+class Graph {
+  private graph = new Map<
+    string,
+    {
+      filename: string;
+      specifier: string;
+    }[]
+  >();
+
+  addEdge(start: string, edge: Edge): void {
+    if (this.graph.has(start)) {
+      this.graph.get(start)?.push(edge);
+    } else {
+      this.graph.set(start, [edge]);
+    }
+  }
+
+  hasEdge(name: string): boolean {
+    return this.graph.has(name);
+  }
+
+  getEdges(name: string): Edge[] {
+    return this.graph.get(name) ?? [];
+  }
```

Comment:
I couldn't understand this graph at a glance. I think it's supposed to be an adjancency list indexed by the file names, and the specifiers are extra data for each edge right? This should be added as comments.

These functions' names are not apparent. I think they should be called `hasNode` and `getOutEdges`.